src/cephadm/design_choices.rst

```diff
@@ -0,0 +1,165 @@
+Cephadm Design Docs
+===================
+
+
+**Housekeeping note:**
+
+If you want to see things added to this document, please add it via Github's `suggestion[Suggest Changes]` feature.
+This allows me to just hit the button in order for the comment to end up in the document.
+
+
+Intro
+-----
+
+Current situation:
+
+Cephadm manages all registered hosts. This means that it needs to periodically scrape data from each
+host to identify changes on the host like:
+
+* disk added/removed
+* daemon added/removed
+* host network/firewall etc has changed
+
+
+Currently, cephadm scrapes each host (up to 10 in parallel[0]) every 6 minutes[1], unless a refresh is forced manually.
+
+Refreshes for disks (ceph-volume), daemons (podman/docker), etc, happen in sequence.
+
+While the refresh model isn't ideal in the first place, the execution time vastly increases with the cluster size > 10
+
+This certainly isn't acceptable since ceph itself scales massively and so has the management part.
+
+There are a couple of solutions to this:
+
+
+1. Pre fetching/caching data on the host
+----------------------------------------
+
+This would allow bin/cephadm to autonomously scrape data that can later be retrieved by mgr/cephadm.
+
+i.e. `cephadm ls` would access an underlying (transparent) cache.
+
+This, btw assumes a **pull** model.
+
+There is at least one possible solutions on *how* to do this.
+
+1.1 daemonize bin/cephadm
+--------------------------
+
+A daemon is able to execute scraping tasks more frequently (as opposed to only when asked by mgr/cephadm)
+
+
+Downsides:
+
+We're still `scp'ing` bin/cephadm from the ceph-mgr to the host which is then executed.
+A daemon on the other hand needs to run continuously on the target system. If ceph-mgr (and the containing bin/cephadm
+script) is updated, the daemon needs to be informed of this event and get updated/restarted as well.
+
+
+1.1.1 HTTP:
+___________
+
+There is a implementation by Paul Cuzner [2] that uses HTTP to serve an endpoint to the hosts metadata.
+
+Upside:
+
+* We could utilize a proper REST API to CRUD all of the daemons/disks etc.
+  Currently it feels a bit weird to call a binary from a remote host (that's another topic though)
+
+
+Downside:
+
+* This is quite some shift in architecture.
+* We still use SSH to bootstrap a node (in order to install bin/cephadm for the first time)
+  Having two distinct transport layers feels odd.
+* The current approach of delivering bin/cephadm to the host doesn't allow the use of external dependencies.
+  This means that we're stuck with the built-in HTTP server lib, which isn't meant for production purposes.
+  bin/cephadm needs to be packaged and distributed (one way or the other) for us to make use of a production ready
+  http server.
+
+Requirements:
+
+* https
+* authentication
+* documentation
+* api versioning (keep it stable)
```

Comment:
why https and authentication? The goal of the PR I have is to provide an exporter - this document is reframing the problem to deliver an agent capability - which is a lot more impactful.

The metadata the PR returns is no more super-secret than the data that mgr/prometheus or the node-exporter exposes - and those endpoints are not https, or authenticated - so it's confusing to me to place these requirements on this PR.

IMO, you only need to consider https/auth if the endpoint is going to become a full agent - with CRUD functionality...and this would be IMPACTFUL, and likely require the re-architecture/packaging of cephadm itself to happen first.

Reply:
We'll likely align the other operations with your approach. Meaning we'd need authentication and https in the long run. So why going with a half-baked solution if we know what we'll need soon.

Or using @jmolmo's metaphor.

"Let's not build that abomination of a bicycle and rather build a proper car"

Reply:
These 2 things would alleviate the fact that `http.server` is not recommended in production:
* **https**: trivial to add to Python `http.server` [[ref]](https://blog.anvileight.com/posts/simple-python-http-server/)
* **authentication**: [same](https://github.com/goya191/SimpleAuthServerSSL.py/blob/master/SimpleHttpsAuthServer.py) (as the authentication source, a key API token read from ceph-mon KV DB would suffice).

The following 2 are not transport specific requirements (they'd apply to every other transport layer):
* **versioning/stability**: I think it's too early for something solely generated and consumed by cephadm. Dashboard had to include versioning in Ceph API because its data came from other component and it was required to keep it stable to third-parties. As this data is fully under cephadm control, it can be added in an additive manner, so it's backward compatible.
* **documenting**: desirable, yeah, but I wouldn't make this a blocker for an internal API.

Reply:
@jschmid1 :) the exporter is only half-backed (no offence taken btw), if you change the scope to make the endpoint become a fully fledged agent. The intent of the PR is to make the data available to orch and UI, more current and scalable - to drive improvements in the UI layers to deliver a better UX.  It adopts the same security principles that we currently apply with mgr/prometheus and node-exporter - so it aligns with what "we" currently define as appropriate.

@jan--f @jschmid1 I get why peoples minds think of how we could bend this to an agent architecture discussion. But that's a choice you're making. We could also choose to look at this as it was intended!